vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoGridFsClientImpl.java

```diff
@@ -0,0 +1,270 @@
+    GridFSDownloadStream stream = bucket.openDownloadStream(objectId);
+    MongoGridFsDownload download = new MongoGridFsDownloadImpl(stream, vertx, config);
+    resultHandler.handle(Future.succeededFuture(download));
+    return this;
+  }
+
+  @Override
+  public MongoGridFsClient downloadFile(String fileName, Handler<AsyncResult<Long>> resultHandler) {
+    requireNonNull(fileName, "fileName cannot be null");
+    requireNonNull(resultHandler, "resultHandler cannot be null");
+
+    return downloadFileAs(fileName, fileName, resultHandler);
+  }
+
+  @Override
+  public MongoGridFsClient downloadFileAs(String fileName, String newFileName, Handler<AsyncResult<Long>> resultHandler) {
+    requireNonNull(fileName, "fileName cannot be null");
+    requireNonNull(newFileName, "newFileName cannot be null");
+    requireNonNull(resultHandler, "resultHandler cannot be null");
+
+    OpenOptions options = new OpenOptions().setWrite(true);
+
+    vertx.fileSystem().open(newFileName, options, asyncFileAsyncResult -> {
+      if (asyncFileAsyncResult.succeeded()) {
+        AsyncFile file = asyncFileAsyncResult.result();
+        GridFSOutputStream gridFSOutputStream = GridFSOutputStream.create(file);
+        bucket.downloadToStream(fileName, gridFSOutputStream, wrapCallback(resultHandler));
+      } else {
+        resultHandler.handle(Future.failedFuture(asyncFileAsyncResult.cause()));
+      }
+    });
+
+    return this;
+  }
+
+  @Override
+  public MongoGridFsClient downloadFileByID(String id, String fileName, Handler<AsyncResult<Long>> resultHandler) {
+    requireNonNull(fileName, "fileName cannot be null");
+    requireNonNull(resultHandler, "resultHandler cannot be null");
+
+    OpenOptions options = new OpenOptions().setWrite(true);
+
+    vertx.fileSystem().open(fileName, options, asyncFileAsyncResult -> {
+      if (asyncFileAsyncResult.succeeded()) {
+        AsyncFile file = asyncFileAsyncResult.result();
+        GridFSOutputStream gridFSOutputStream = GridFSOutputStream.create(file);
+        ObjectId objectId = new ObjectId(id);
+        bucket.downloadToStream(objectId, gridFSOutputStream, wrapCallback(resultHandler));
+      } else {
+        resultHandler.handle(Future.failedFuture(asyncFileAsyncResult.cause()));
+      }
+    });
+
+    return this;
+  }
+
+  @Override
+  public MongoGridFsClient drop(Handler<AsyncResult<Void>> resultHandler) {
+    requireNonNull(resultHandler, "resultHandler cannot be null");
+
+    bucket.drop(wrapCallback(resultHandler));
+    return this;
+  }
+
+  @Override
+  public MongoGridFsClient findAllIds(Handler<AsyncResult<List<String>>> resultHandler) {
+    requireNonNull(resultHandler, "resultHandler cannot be null");
+
+    List<String> ids = new ArrayList<>();
+
+    Context context = vertx.getOrCreateContext();
+    bucket.find()
+      .forEach(gridFSFile -> {
+          ids.add(gridFSFile.getObjectId().toHexString());
+        },
+        (result, throwable) -> {
```

Comment:
ideally the ids list should be copied here because the thread of the find callback might not be the same than the context thread